target.go

```diff
@@ -0,0 +1,208 @@
+package flashlight
+
+import (
+	"sync"
+
+	"google.golang.org/grpc"
+)
+
+// Target represents some dialable gRPC target, typically a single gRPC server.
+type Target struct {
+	// Group is a driver-defined identifier for a group of targets.
+	//
+	// Targets within the same (non-empty) group are considered interchangeable.
+	Group string
+
+	// Name is the target name used to dial the endpoint. The syntax is defined
+	// in https://github.com/grpc/grpc/blob/master/doc/naming.md.
+	Name string
+
+	// Options is a set of grpc.DialOptions used when dialing this target.
+	Options []grpc.DialOption
+
+	// MetaData contains driver-specific meta-data about the target.
+	MetaData interface{}
+}
+
+// A TargetObserver is notified when targets are added or removed from a
+// TargetRegistry.
+type TargetObserver interface {
+	// TargetAdded is called when a target is added to an observed registry.
+	TargetAdded(*Target)
+
+	// TargetRemoved is called when a target is removed from an observed
+	// registry.
+	TargetRemoved(*Target)
+}
+
+// TargetRegistry is a container for discovered gRPC servers.
+type TargetRegistry struct {
+	m         sync.RWMutex
+	targets   map[*Target]struct{}
+	observers map[TargetObserver]struct{}
+	groups    map[string]map[*Target]struct{}
+}
+
+// Add adds a target to the registry.
+func (r *TargetRegistry) Add(t *Target) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	if r.targets == nil {
+		r.targets = map[*Target]struct{}{}
+	} else if _, ok := r.targets[t]; ok {
+		return
+	}
+
+	r.targets[t] = struct{}{}
+
+	if t.Group != "" {
+		if r.groups == nil {
+			r.groups = map[string]map[*Target]struct{}{}
+		}
+
+		g, ok := r.groups[t.Group]
+
+		if !ok {
+			g = map[*Target]struct{}{}
+			r.groups[t.Group] = g
+		}
+
+		g[t] = struct{}{}
+	}
+
+	r.notifyAll(TargetObserver.TargetAdded, t)
+}
+
+// Remove removes a target from the registry.
+func (r *TargetRegistry) Remove(t *Target) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	if _, ok := r.targets[t]; !ok {
+		return
+	}
+
+	delete(r.targets, t)
+	delete(r.groups[t.Group], t)
+
+	r.notifyAll(TargetObserver.TargetRemoved, t)
+}
+
+// Has returns true if t is in the registry.
+func (r *TargetRegistry) Has(t *Target) bool {
+	r.m.RLock()
+	defer r.m.RUnlock()
+
+	_, ok := r.targets[t]
+	return ok
+}
+
+// Select returns a random target in the given group.
+//
+// If there are targets in the given group, ok is false.
```

Comment:
I think you mean if there are **no** targets then ok is false?
```suggestion
// If there are no targets in the given group, ok is false.
```